drivers/soundwire/intel.c

```diff
@@ -750,6 +817,12 @@ static int intel_hw_params(struct snd_pcm_substream *substream,
 	int ret, i, ch, dir;
 	bool pcm = true;
 
+	ret = sdw_stream_setup(substream, dai);
```

Comment:
Why do we do sdw_stream_step() in hw_params().I think it should be done in startup() to keep it in line with what is there already

Reply:
I don't think it should be done in a dai operation, the notion of stream is larger than a dai.
For example the same stereo 'stream' can be rendered by using link0-dai0 and link1-dai4, going to two separate amplifiers. This setup should be done in a dailink operation where we will have the list of all cpu and codec dais handled.

Reply:
I am still struggling here, in the initial proposal there was no notion of DAI, 

Initially we did all the soundwire stream handling in pcm_ops callbacks.

Here everything is done as part of dai_ops callbacks.

That's very very different!

Reply:
@plbossart but all the dai_ops are called from the pcm ops context anyway isnt it? like for example: 
https://github.com/thesofproject/linux/blob/b45315d57e2ab6fce1de8bd6391d925143b4fbda/sound/soc/soc-pcm.c#L806

Reply:
Sorry @ranj063, I don't get your point. I never fully understood how DPCM works so not sure if the two models are was referring to are equivalent.

Does anyone have a sequence diagram showing how those rtd, pcm_ops and  cpu_dai and codec_dai ops are called?

Reply:
I will check the ASOC and give a process flow picture

Reply:
> I don't think it should be done in a dai operation, the notion of stream is larger than a dai.
> For example the same stereo 'stream' can be rendered by using link0-dai0 and link1-dai4, going to two separate amplifiers. This setup should be done in a dailink operation where we will have the list of all cpu and codec dais handled.

I agree with @plbossart the operation can be done in dailink.  Actually I think intel_hw_params() will be invoked twice if there are two sdw components in the same dailink. @RanderWang Can we allocate and release stream in machine driver? Maybe we can use startup and shutdown ops?

Reply:
![Drawing1](https://user-images.githubusercontent.com/39945651/63092883-cc4ad780-bf95-11e9-854c-c4e676bfc1f0.png)

Reply:
There are component driver and dai driver in ASoC.  they all have prepare, hw_params, and trigger functions. In the following picture,   dai hw_params is for dai driver , and  the only hw_params is for component driver. Now there is only component driver in SOF for SDW, 
and there is only dai driver in SDW component and codec comopnent.

All the components driver are invoked in both FE and BE dai, so you will find some functions are invoke twice. Although there is no component driver in SDW and codec component, I still
marked some functions in in picture to reflect the process flow in ASoC

![Drawing2](https://user-images.githubusercontent.com/39945651/63093471-fb624880-bf97-11e9-9b66-ecd4dfd5270e.png)